components/core/src/clp/hash_utils.hpp

```diff
@@ -0,0 +1,122 @@
+#ifndef CLP_HASH_UTILS_HPP
+#define CLP_HASH_UTILS_HPP
+
+#include <cstddef>
+#include <span>
+#include <string>
+#include <vector>
+
+#include <openssl/evp.h>
+
+#include "ErrorCode.hpp"
+#include "TraceableException.hpp"
+
+namespace clp {
+/**
+ * A C++ wrapper for openssl's EVP digest message digest context (EVP_MD_CTX).
+ */
+class EvpDigestContext {
+public:
+    // Types
+    class OperationFailed : public TraceableException {
+    public:
+        // Constructors
+        OperationFailed(ErrorCode error_code, char const* const filename, int line_number)
+                : TraceableException(error_code, filename, line_number) {}
+
+        // Methods
+        [[nodiscard]] auto what() const noexcept -> char const* override {
+            return "EvpCtxManager operation failed";
+        }
+    };
+
+    // Constructors
+    /**
+     * Creates and initializes a hash context by calling EVP_MD_CTX_create
+     * and EVP_DigestInit_ex from OpenSSL library
+     * @param type the type of digest (hash algorithm) from OpenSSL library.
+     * @throw EvpCtxManager::ErrorCode_Failure if EVP_MD_CTX_create fails.
+     * @throw EvpCtxManager::ErrorCode_NotInit if EVP_DigestInit_ex fails.
+     */
+    EvpDigestContext(EVP_MD const* type) : m_md_ctx{EVP_MD_CTX_create()} {
+        if (nullptr == m_md_ctx) {
+            throw OperationFailed(ErrorCode_Failure, __FILENAME__, __LINE__);
```

Comment:
Can we use [ERR_get_error](https://docs.openssl.org/1.0.2/man3/ERR_get_error/) and [ERR_error_string](https://docs.openssl.org/1.0.2/man3/ERR_error_string/) to get and return more info about the error in the exceptions below? Then we wouldn't really have to specify which method failed in the docstring above.